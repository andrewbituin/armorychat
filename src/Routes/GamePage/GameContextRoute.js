import React from 'react';
import { GameProvider } from "../../Contexts/gameContext";
import GamePage from './GamePage';

export default function GameContextRoute(props) {
  return (
    <GameProvider>
      <GamePage {...props}/>
    </GameProvider>
  )
}