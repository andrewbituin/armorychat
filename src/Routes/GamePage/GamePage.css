* {
  box-sizing: border-box;
}

.squad-details {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  position: fixed;
  background-color: #2c2e35;
  top: 0;
  left: 90px;
  width: 380px;
  height: 100%;
  border-right: 1px solid #6e6f72;
  border-left: 1px solid #6e6f72;
}

.squad-details-top {
  padding: 20px;
}

.squad-details__main {
  padding: 10px;
  margin-bottom: 60px;
}

.squad-details__main h2 {
  font-size: 32px;
  font-weight: bold;
}

.squad-details__image {
  position: absolute;
  opacity: 0.6;
  left: 0;
  top: -109px;
  z-index: -1;
  width: 100%;
  clip-path: inset(109px 0 109px 0);
  overflow: hidden;
}

.squad-details::after {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  content: ' ';
  height: 160px;
  z-index: -1;
  background: linear-gradient(#00000000, #2c2e35);
}

.small-detail {
  background-color: #3b3f48;
  font-size: 14px;
  border-radius: 5px;
  box-shadow: 0 1px 8px 0 #00000080;
  border: 1px solid #5b6370;
  padding: 3px 6px;
  margin-right: 10px;
  color: #fff;
}

.squad-details__create-container {
  display: flex;
  left: 0;
  right: 0;
  align-items: center;
  justify-content: center;
  background-color: #27282e;
  height: 15%;
  min-height: 100px;
  bottom: 0;
}

.green-button, .grey-button {
  display: flex;
  justify-content: center;
  align-items: center;
  outline: none;
  border: none;
  border-radius: 5px;
  box-shadow: 0 1px 8px 0 #00000080;
  font-family: Rajdhani, sans-serif;
  font-weight: 600;
  color: #fff;
  font-size: 16px;
}

.green-button:disabled {
  color: #a3a3a3;
  background-image: linear-gradient(#6e6e6e, #444444);
}

.green-button{
  background-image: linear-gradient(#5a801f, #406406);
}

.grey-button{
  background-image: linear-gradient(#586166,#3b4449);
}

.green-button:hover:not(:disabled), .green-button:focus {
  background-image: linear-gradient(#6fa42c, #538810);
}

.grey-button:hover, .grey-button:focus {
  background-image: linear-gradient(#60696d,#444e52);
}

.green-button:hover, .grey-button:hover {
  cursor: pointer;
}

.create-squad-button{
  width: 80%;
  height: 36px;
}

.squad-list-container {
  position: relative;
  margin-left: 470px;
}

.green-button-flat, .green-button-flat__selected {
  font-family: Rajdhani, sans-serif;
  min-width: 45px;
  padding: 0 12px;
  color: #fff;
  border-radius: 5px;
  margin-right: 10px;
  font-size: 14px;
  font-weight: 500;
  height: 30px;
  margin-bottom: 5px;
}

.green-button-flat {
  background-color: #43582d;
  border: none;
  outline: none;
}

.green-button-flat:hover {
  background-color: #6fa42c;
  cursor: pointer;
}

.green-button-flat__selected {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background-color: #5a801f;
}

.squads-list__top {
  position: fixed;
  top: 0;
  padding-top: 20px;
  margin: 0 10px;
  width: calc(100% - 490px);
  background-color: #212227;
  border: 4px solid #212227;
  z-index: 2;
}

.squads-list__top-titles {
  list-style: none;
  padding: 0;
  padding-bottom: 18px;
  margin-bottom: 0;
  display: flex;
  border-image-source:url('../../images/border-image.png');
  border-image-width: 0 0 1px 0;
  border-image-slice: 1;
  border-image-repeat: repeat;
  border-bottom: 1px dashed #565d69;
}

.squads-list__top-titles li {
  display: inline-flex;
  justify-content: center;
  color: #565d69;
  text-transform: uppercase;
}

.squads-list__top-titles__info {
  width: 261px;
}

.squads-list__top-titles__spots {
  width: 320px;
}

.squads-list__top-titles__owner {
  width: 200px;
}

.squads-list__top-titles__requirements {
  text-align: right;
  margin-left: auto;
  margin-right: 40px;
}

.show-squad-details-button {
  height: 30px;
  width: 30px;
}

.squad-list__top-gamemode-buttons {
  display: flex;
  align-items: center;
}

.squad-list__top-dropdown-container {
  width: 250px;
  display: inline-block;
  margin: 4px 5px;
}

.squad-list__top-dropdown-container .r-dropdown__control, .squad-list__top-dropdown-container .dropdown-result-button {
  margin: 0 5px;
}

.gamemode-dropdown .r-dropdown__control:hover {
  background-color: #141416;
}

.active-squad {
  width: 300px;
  height: 200px;
}

.show-squad-details-button {
  display: none;
}

@media screen and (max-width: 1400px) {
  .squad-details {
    width: 300px;
  }

  .squad-list-container {
    margin-left: 390px;
  }

  .squads-list__top {
    width: calc(100% - 410px);
  }

  .squad-list h4 {
    font-size: 16px;
    min-width: 120px;
  }

  .squad__owner {
    width: 140px;
  }

  .squad__owner p {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  .squad-details__image {
    clip-path: inset(29px 0 29px 0);
  }
}

@media screen and (max-height: 780px), (max-width: 1280px) {
  .squad-details {
    display: none;
  }

  .squad-list-container {
    margin-left: 90px;
  }

  .squads-list__top {
    width: calc(100% - 110px);
  }

  .display__game-info {
    display: flex !important;
  }

  .show-squad-details-button {
    display: inline-block;
  }

  .squad-details {
    position: relative;
    top: 124px;
    height: unset;
    left: 90px;
    width: calc(100% - 120px);
    margin: 0 15px;
    border: none;
    background-color: transparent;
    border-bottom: 4px solid #212227;
  }

  .squad-details__create-container {
    border-radius: 5px;
    border-image-source:url('../../images/border-image.png');
    border-image-width: 0 0 1px 0;
    border-image-slice: 1;
    border-image-repeat: repeat;
    border-bottom: 1px dashed #565d69;
    margin-bottom: 10px;
  }

  .squad-details::after {
    display: none;
  }

  .squad-details__image, .squad-details__main {
    display: none;
  }
}

@media screen and (max-width: 950px) {
  .squad__owner {
    display: none;
  }

  .squads-list__top-titles__owner {
    display: none !important;
  }
}


@media screen and (max-width: 900px) {
  .squad-list {
    padding: 0 20px;
  }

  .squad-container {
    margin-bottom: 10px;
  }

  .squad-list-container {
    margin-left: 0;
  }

  .squads-list__top-titles li {
    display: none;
  }

  /* .squads-list__top-titles {
    display: none;
  } */

  .squads-list__top {
    position: static;
    margin-top: -10px;
    width: calc(100% - 20px);
  }

  .squad-list {
    margin-top: 20px;
  }

  .squad-container {
    height: 118px;
    padding-bottom: 50px;
  }

  .squad__spots-list {
    position: absolute;
    bottom: 0;
    left: 10px;
  }

  .squad__details {
    width: calc(100% - 60px);
  }

  .squad__requirements {
    position: absolute;
    right: 0;
    top: 2px;
  }

  .squad__requirements .small-detail {
    margin-right: 5px;
  }

  .squad__gamemode-image-container {
    padding: 0;
    display: inline-flex;
    margin-right: 10px;
  }

  .squad-details {
    position: absolute;
    top: 192px;
    left: 0;
    width: calc(100% - 30px);
    z-index: 2;
  }
}

@media screen and (max-width: 400px) {
  .squad__requirements {
    display: none;
  }
}