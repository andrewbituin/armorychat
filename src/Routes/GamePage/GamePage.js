import React, { useState, useEffect, useContext, useRef } from "react";
import config from "../../config";
import TokenService from "../../services/token-service";
import GameContext from '../../Contexts/gameContext';
import helpers from '../../services/helpers';
import Squad from '../../Components/Squad/Squad';
import Error from '../../Components/Error/Error';
import GameInfo from '../../Components/GameInfo/GameInfo';

import './GamePage.css';

import io from 'socket.io-client';
import Dropdown from "../../Components/Dropdown/Dropdown";
import Modal from "../../Components/Modal/Modal";
import ActiveSquad from '../../Components/ActiveSquad/ActiveSquad';
let socket;

export default function GamePage(props) {
  const gameContext = useContext(GameContext);
  const [loading, toggleLoading] = useState(true);
  const [activeSquad, setActiveSquad] = useState(undefined);
  const squadList = useRef(null);

  function getGame() {
    return fetch(`${config.API_ENDPOINT}${props.match.url}`, {
      headers: {
        authorization: `bearer ${TokenService.getAuthToken()}`
      },
      body: JSON.stringify(TokenService.getAuthToken)
    })
      .then(res =>
        (!res.ok)
          ? res.json().then(e => Promise.reject(e))
          : res.json()
      )
      .then(game => {
        gameContext.setGame(game);
      });
  }

  async function populateContext() {
    try {
      await getGame();
      await gameContext.getAllParties();
      toggleLoading(false);
    } catch(err) {
      gameContext.setError(err);
    }
  }

  //throws an error - fix later
  // useEffect(() => {
  //   browserBack()
  // },[])

  // function browserBack(){
  //   window.history.pushState(null, null, '/');
  //   window.onpopstate = function () {
  //       window.history.go(1);
  //   };
  // }

  useEffect(() => {
    // populate party listing from API
    populateContext();
    // get game stuff
    // connect to socket io for this game
    socket = io('https://limitless-brushlands-45977.herokuapp.com');
    socket.emit('join room', props.match.url);

    return () => {
      socket.emit('leave game');
      socket.disconnect();
    };
  }, []);

  useEffect(() => {
    document.addEventListener('scroll', handlePartiesScroll);

    socket.on('posted party', async function(data) { 
      let pagesAvailable = data.pages_available;
      let partiesAvailable = data.parties_available;
      //if we have filters on we need an accurate count
      if (checkIfFilters()) {
        const temp = await gameContext.getAllParties(true);
        pagesAvailable = temp.pages_available;
        partiesAvailable = temp.parties_available;
      }
      if (checkFilters(pagesAvailable, data.party)) {
        gameContext.setParties([...gameContext.parties, {
          ...data.party,
          pages_available: pagesAvailable,
          parties_available: partiesAvailable
        }]);
      }
    });

    socket.on('spot updated', function(data) {
      let index = helpers.findPartyIndex(data.party.id, gameContext.parties);
      if (index !== -1) {
        let newParties = [...gameContext.parties];
        newParties[index] = data.party;
        gameContext.setParties(newParties, data.pages_available);
      }
    });

    socket.on('delist party', function(data) {
      let index = helpers.findPartyIndex(data.party.id, gameContext.parties);
      if (index !== -1) {
        let newParties = [...gameContext.parties];
        newParties.splice(index, 1);
        gameContext.setParties(
          newParties, 
          data.pages_available,
          (data.pages_available < gameContext.currentPage) && data.pages_available
        );
      }
    })

    return () => {
      socket.off('posted party');
      socket.off('spot updated');
      document.removeEventListener('scroll', handlePartiesScroll);
    }
  }, [gameContext.parties]);

  function checkIfFilters() {
    return (!!gameContext.searchTerm || !!gameContext.gamemodeFilter || gameContext.requirementFilters.length > 0 || gameContext.roleFilters.length > 0) ;
  }

  function checkFilters(pagesAvailable, party) {
    const { requirementFilters, roleFilters, searchTerm, gamemodeFilter, requirements, roles, gamemodes } = gameContext;
    if (gameContext.currentPage !== (pagesAvailable - 1)) return false;
    let relevant = true;

    //requirement filters check
    if (requirementFilters.length > 0) {
      for (let i = 0; i < requirementFilters.length; i++) {
        let temp = false;
        for (let j = 0; j < party.reqs.length; j++) {
          if (party.reqs[j].name === requirements[requirementFilters[i]].name) temp = true;
        }
        if (temp === false) relevant = false;
        break;
      }
    }
    if (relevant === false) return false;

    //role filters check
    if (roleFilters.length > 0) {
      //for each filter
      for (let i = 0; i < roleFilters.length; i++) {
        let temp = false;
        //for each spot
        for (let j = 0; j < party.spots.length; j++) {
          //check for each role in each spot
          for (let k = 0; k < party.spots[j].roles.length; k++) {
            if (party.spots[j].roles[k].name === roles[roleFilters[i]].name) temp = true;
          }
        }
        if (temp === false) relevant = false;
      }
    }
    if (relevant === false) return false;

    //searchTerm check
    if (searchTerm) {
      if (!(party.title.includes(searchTerm) || party.description.includes(searchTerm))) relevant = false;
    }
    if (relevant === false) return false;

    //gamemode check
    if (gamemodeFilter) {
      if (party.gamemode.name !== gamemodes[gamemodeFilter].name) return false;
    }

    return relevant;
  }

  function handleSquadOnClick(index) {
    setActiveSquad(index);
  }

  function generateParties() {
    if (gameContext.parties.length < 1) {
      return <p>No parties available...</p>;
    }
    return gameContext.parties.map((party, index) => (
      <Squad key={index} index={index} party={party} gameId={props.match.params.gameId} onClick={handleSquadOnClick}/>
    ));
  }

  async function handlePartiesScroll() {
    //document, NOT window
    if (document.documentElement.scrollTop/(document.documentElement.scrollHeight - document.documentElement.clientHeight) === 1) {
      if (gameContext.currentPage < gameContext.pagesAvailable - 1) {
        gameContext.incrementCurrentPage(gameContext.getAllParties);
      }
    };
  }

  function generateGamemodeDropdown() {
    return <div className="squad-list__top-dropdown-container">
      <Dropdown
        name="gamemode"
        className='gamemode-dropdown'
        active={gameContext.gamemodeFilter}
        gameId={gameContext.id}
        onChange={e => gameContext.setGamemodeFilter(e.value)}
        onButtonClick={e => gameContext.setGamemodeFilter(0)}
        placeholder='All'
        startValue={gameContext.gamemodeFilter}
        options={gameContext.gamemodes}
      />
    </div>
  }

  if (loading) {
    return <div>Loading...</div>
  }
  return (
    <div className="container">
      {gameContext.error && <Error close={gameContext.clearError} error={gameContext.error}/>}
      <GameInfo history={props.history} roomUrl={props.match.url}/>
      <div className="squad-list-container">
        <div className="squads-list__top">
          <div className="squad-list__top-gamemode-buttons">
            {generateGamemodeDropdown()}
            <button 
              className="show-squad-details-button green-button" 
              type="button" aria-label="show options" 
              onClick={() => {
                //hacky... change this later! Also works poorly with resizing while menu is open.
                let classList = document.getElementById('game-info').classList;
                classList.toggle('display__game-info');
                if (classList.contains('display__game-info') && document.documentElement.clientWidth < 1267) {
                  if (document.documentElement.clientWidth < 883) {
                    squadList.current.setAttribute('style', 'margin-top: 586px;');
                  } else {
                    squadList.current.setAttribute('style', 'margin-top: 120px;');
                  }
                } else {
                  squadList.current.setAttribute('style', 'margin-top: 20px;');
                }
              }}
            >
              <i className="fas fa-bars"/>
            </button>
          </div>
          <ul className="squads-list__top-titles">
            <li className="squads-list__top-titles__info">Squad Info</li>
            <li className="squads-list__top-titles__spots">Spots</li>
            <li className="squads-list__top-titles__owner">Owner</li>
            <li className="squads-list__top-titles__requirements">Requirements</li>
          </ul>
        </div>
        {gameContext.partiesLoading 
          ? (
              <div className="squad-list">
                Loading...
              </div>
            )
          : (
              <ul className="squad-list" ref={squadList}>
                {generateParties()}
              </ul>
            )}
        <Modal
          isShowing={activeSquad || activeSquad===0}
          hide={() => setActiveSquad(undefined)}
          content={
            (gameContext.parties[activeSquad] 
              ? <ActiveSquad party={gameContext.parties[activeSquad]} hide={() => setActiveSquad(undefined)} history={props.history}/>
              : null
            )
          }
        />
      </div>
    </div>
  );
}
