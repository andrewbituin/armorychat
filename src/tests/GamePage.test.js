import React from 'react';
import GamePage from '../Routes/GamePage/GamePage';
import { MemoryRouter } from 'react-router-dom';
import renderer from "react-test-renderer";
import { shallow } from 'enzyme';

describe('GamePage component', () => {

  it('renders without crashing', () => {
    shallow(<GamePage />)
  })

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <GamePage />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})