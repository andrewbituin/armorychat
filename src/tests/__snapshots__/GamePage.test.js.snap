// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePage component renders the UI as expected 1`] = `
<div>
  Loading...
</div>
`;
